### PR TITLE
Listing stacks

### LINKED ISSUE #2410: List recent deployments of a service in the CLI
# This is a Feature Proposal

## Description

To understand what recent deployments have been made users should be able to see recent deployments made for this particular service. This is also helpful when trying to roll back so before running the rollback command users can inspect what has been deployed before.

`serverless deploy list`

Similar or dependent issues:
* #12345

## Additional Data

* ***Serverless Framework Version you're using***:
* ***Operating System***:
* ***Stack Trace***:
* ***Provider Error messages***:

### PR DESCRIPTION
**Pull request by nikgraf**
## What did you implement:

Closes #2410

## How did you implement it:

Fetching the S3 Objects from the deployment folder based on the default stage & region. In case no entries are found the following message is shown:

```
Couldn\'t find any existing stacks.
Please verify that stage and region are correct.
```

This uses the same findAndGroupStacks helper functions as in this PR https://github.com/serverless/serverless/pull/2436. Since it's exactly the same file it shouldn't even create a merge conflict.

## Preview

screen shot 2016-10-18 at 15 41 32

## How can we verify it:

1. Deploy a service and run `serverless deploy list`
2. You should see your deployment listed.

## Questions

@flomotlik I believe it would be useful in case `deploy list` would accept `stage` and `region` as parameters. If yes, should I add it in this PR or another?
Please also check the preview if it the output is consistent with other commands.

***Is this ready for review?:*** YES

### CODE REVIEW

**Comment by flomotlik** [comment]
@nikgraf shouldn't add stage and region in here, but can use the stage and region config in serverless.yml. We have to reimplement stage and region handling from options at some point.

**Comment by nikgraf** [comment]
@flomotlik @eahefnawy @pmuens ready to be merged from my POV

**Comment by flomotlik** [comment]
GTM from my side. @serverless/core could one of you check out the code and validate (please assign yourself if you take the task). Then it can be merged.

File: lib/plugins/aws/deploy/lib/listStacks.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+const BbPromise = require('bluebird');
+const findAndGroupStacks = require('../utils/findAndGroupStacks');
+
+module.exports = {
+  listStacks() {
+    const service = this.serverless.service.service;
+    const stage = this.options.stage;
+
+    return this.sdk.request('S3',
```
Comment (pmuens) [inline lib/plugins/aws/deploy/lib/listStacks.js]: Could you replace this with a usage of the recently introduced `AwsProvider` plugin usage (Then it should be something like `this.provider.request`).

Here's how @eahefnawy has done it in the custom resource merging PR fix:
https://github.com/serverless/serverless/pull/2385/commits/f0804d244f2fbb353a944468225d17f663c5d916

File: lib/plugins/deploy/deploy.js [hunk 2]
```diff
@@ -56,6 +56,12 @@ class Deploy {
               },
             },
           },
+          list: {
```
Comment (pmuens) [inline lib/plugins/deploy/deploy.js]: Could you add the corresponding command to the `AwsDeploy` plugin as well?

Comment (pmuens) [inline lib/plugins/deploy/deploy.js]: Ah my bad. This is only necessary for `options`.

**Comment by pmuens** [comment]
Yes. That's right. Basically it should be a separate plugin (`deployList`) as it uses a nested command and should be implemented the same way the `deployFunction` plugin is implemented.

File: lib/plugins/aws/utils/findAndGroupStacks.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+'use strict';
```
Comment (flomotlik) [inline lib/plugins/aws/utils/findAndGroupStacks.js]: can you rename the file to `findAndGroupDeployments.js`

File: docs/03-cli-reference/03-deploy.md [hunk 4]
```diff
@@ -41,6 +41,12 @@ serverless deploy --stage production --region eu-central-1
 With this example we've defined that we want our service to be deployed to the `production` stage in the region
 `eu-central-1`.
 
+## List existing deploys
```
Comment (flomotlik) [inline docs/03-cli-reference/03-deploy.md]: Explanation text please, e.g. Something like "List recent deployments available in your S3 Deployment Bucket". Also mention to use the stage/region config in the provider config element.

File: lib/plugins/aws/utils/findAndGroupStacks.js (hunk 3 above)
Comment (nikgraf) [inline lib/plugins/aws/utils/findAndGroupStacks.js]: done, btw we use stack a lot e.g. createStack. What's the difference between a stack and a deployment from your pov?

File: docs/03-cli-reference/03-deploy.md (hunk 4 above)
Comment (nikgraf) [inline docs/03-cli-reference/03-deploy.md]: done 👍

**Comment by flomotlik** [comment]
GTM from my side, @pmuens if your reviews are fine as well lets get this out

File: lib/plugins/aws/deployList/tests/index.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
+'use strict';
+
+const sinon = require('sinon');
+const BbPromise = require('bluebird');
+const expect = require('chai').expect;
+const AwsDeployList = require('../index');
+const AwsProvider = require('../../provider/awsProvider');
+const Serverless = require('../../../../Serverless');
+
+describe('listDeployments', () => {
```
Comment (pmuens) [inline lib/plugins/aws/deployList/tests/index.js]: We've used the name of the plugin here --> `AwsDeployList` in other plugin related tests (https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deployFunction/tests/index.js#L14). Not sure if this is best practice though.

File: lib/plugins/deploy/deploy.js [hunk 6]
```diff
@@ -56,6 +56,12 @@ class Deploy {
...
+            usage: 'List deployed version of you Serverless Service',
```
Comment (pmuens) [inline lib/plugins/deploy/deploy.js]: `you` --> `your`

File: lib/plugins/aws/deployList/tests/index.js (hunk 5 above)
Comment (nikgraf) [inline lib/plugins/aws/deployList/tests/index.js]: you are right!

File: lib/plugins/deploy/deploy.js (hunk 6 above)
Comment (nikgraf) [inline lib/plugins/deploy/deploy.js]: thx, ❤️ you attention to detail

**Pull request merged**